### PR TITLE
Storage utils

### PR DESCRIPTION
**Pull request by jasonwyatt**
At first glance, this looks like an absolutely massive PR, but a lot of the changed lines are updating indentation. Hopefully "hide whitespace changes" works so you can skip reviewing indentation. 🤞

This is a preamble to the implementation of `ReferenceModeStore`, coming soon. Within this PR I implement several utilities which the `ReferenceModeStore` will use:

* Message - Defines the different types of messages which the `ReferenceModeStore` will send around to the backing store, container store, and storage proxy. When a message is enqueued into the `MessageQueue`, it is associated with a `CompletableDeferred` that will resolve when it's been dequeued.
* MessageQueue - Thread-safe implementation of the `receiveQueue` from `reference-mode-store.ts`.
* RefModeCrdtTypes.kt - The `ReferenceModeStore`, being a *store*, needs to support managing `ProxyMessage`s which contain CrdtModel/CrdtOperation changes. To that end, this file defines some Crdt-esque classes so that users of ReferenceModeStore can treat the managed singleton/set as a Crdt.
* Reference - The small `Referencable` which will be managed by the `containerStore` component of the `ReferenceModeStore`.
* ProxyCallbackManager - A thread-safe manager of `ProxyCallback`s, as the name would imply 😄 
* SendQueue - Thread safe implementation of reference-mode-store.ts's `enqueueSend`, `enqueueBlockingSend`, etc. behavior.

Also, some tweaks were made to existing files:

* RawEntity.kt - Made `RawEntity` implement `Referencable`, so it can be used within the `RefModeStoreData`/`RefModeStoreOp` objects.
* CrdtEntity.kt - Updated the `CrdtEntity.Data` constructors to support more flexible creation from `RawEntity` objects.
* CrdtSet.kt - Mostly indentation, but it does `open` up the `Operation` classes, so they can be extended by the `RefModeStoreOp` classes.
* CrdtSingleton.kt - Opens up the `Operation` classes, so they can be extended by `RefModeStoreOp` classes.
* HoldQueue.kt - Makes the `enqueue` method a suspend function because of the other changes required to make HoldQueue actually thread-safe.

Edit: I guess it's still a pretty large PR.

### CODE REVIEW

File: src_kt/java/arcs/storage/BUILD [hunk 1]
```diff
@@ -4,11 +4,44 @@ load("//third_party/java/arcs/build_defs:build_defs.bzl", "kt_jvm_and_js_library
...
+
+kt_jvm_and_js_library(
+    name = "storage_key",
+    srcs = [
+        "StorageKey.kt",
+    ],
+)
+
+kt_jvm_and_js_library(
+    name = "proxy_stuff",
```
Comment (csilvestrini) [inline src_kt/java/arcs/storage/BUILD]: Just "proxy"?

File: src_kt/java/arcs/storage/util/ProxyCallbackManager.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+/** Thread-safe manager of a collection of [ProxyCallback]s. */
+class ProxyCallbackManager<Data : CrdtData, Op : CrdtOperation, ConsumerData> {
+    private val mutex = Mutex()
+    internal var nextCallbackToken = 1
+    internal val callbacks = mutableMapOf<Int, ProxyCallback<Data, Op, ConsumerData>>()
+
+    /** Adds a [ProxyCallback] to the collection, and returns its token. */
+    fun register(proxyCallback: ProxyCallback<Data, Op, ConsumerData>): Int {
+        while (!mutex.tryLock()) { /* Wait. */ }
```
Comment (csilvestrini) [inline src_kt/java/arcs/storage/util/ProxyCallbackManager.kt]: Is this really the best way of doing this? Does Kotlin have a thread-safe map implementation you could use instead?

File: src_kt/java/arcs/storage/util/SendQueue.kt [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+import arcs.common.ReferenceId
+import arcs.crdt.internal.VersionMap
+import arcs.storage.referencemode.Reference
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * Thread-safe queue to manage and dispatch sending events, where said dispatching may depend on
+ * waiting for a hold in a [HoldQuuee] to be released.
```
Comment (csilvestrini) [inline src_kt/java/arcs/storage/util/SendQueue.kt]: HoldQueue

File: src_kt/java/arcs/storage/util/ProxyCallbackManager.kt (hunk 2 above)
Comment (jasonwyatt) [reply src_kt/java/arcs/storage/util/ProxyCallbackManager.kt]: It doesn't have one without the JVM, unfortunately... I think another way to do a lot of this thread safety would be to use Actors, but I don't feel comfortable with them just yet to make that call-- and I think most engineers understand mutexing.

Comment (csilvestrini) [reply src_kt/java/arcs/storage/util/ProxyCallbackManager.kt]: Ah of course. Thanks for explaining! Yeah the mutex is fine, I just thought there were safer/more sophisticated tools available, but I guess not in kotlin native.

**Pull request merged**